src/nxt_http_static.c

```diff
@@ -235,42 +279,58 @@ nxt_http_static_iterate(nxt_task_t *task, nxt_http_request_t *r,
         nxt_str_set(&chr, "");
     }
 
-    nxt_debug(task, "http static: \"%V\" (chroot: \"%V\")", &shr, &chr);
+    nxt_debug(task, "http static: \"%V\", index: \"%V\" (chroot: \"%V\")",
+              &shr, &idx, &chr);
 #else
-    nxt_debug(task, "http static: \"%V\"", &shr);
+    nxt_debug(task, "http static: \"%V\", index: \"%V\"", &shr, &idx);
 #endif
 #endif /* NXT_DEBUG */
 
-    if (share->is_const) {
+    shr_is_const = share->is_const;
+    idx_is_const = nxt_var_is_const(index);
+
+    if (!shr_is_const || !idx_is_const) {
+        ret = nxt_var_query_init(&r->var_query, r, r->mem_pool);
+        if (nxt_slow_path(ret != NXT_OK)) {
+            nxt_http_request_error(task, r, NXT_HTTP_INTERNAL_SERVER_ERROR);
+            return;
+        }
+    }
+
+    if (idx_is_const) {
+        nxt_var_raw(index, &ctx->index);
+
+    } else {
+        nxt_var_query(task, r->var_query, index, &ctx->index);
+    }
+
+    if (shr_is_const) {
         nxt_var_raw(share->var, &ctx->share);
 
 #if (NXT_HAVE_OPENAT2)
-        if (conf->chroot != NULL && ctx->index == 0) {
+        if (conf->chroot != NULL && ctx->share_idx == 0) {
             nxt_var_raw(conf->chroot, &ctx->chroot);
         }
 #endif
 
-        nxt_http_static_send_ready(task, r, ctx);
-
     } else {
-        ret = nxt_var_query_init(&r->var_query, r, r->mem_pool);
-        if (nxt_slow_path(ret != NXT_OK)) {
-            nxt_http_request_error(task, r, NXT_HTTP_INTERNAL_SERVER_ERROR);
-            return;
-        }
-
         nxt_var_query(task, r->var_query, share->var, &ctx->share);
 
 #if (NXT_HAVE_OPENAT2)
-        if (conf->chroot != NULL && ctx->index == 0) {
+        if (conf->chroot != NULL && ctx->share_idx == 0) {
```

Comment:
There is an optimization here. For example:
```
{
     "chroot": "text$var",
     "share": [...]
}
```
The variable of chroot only resolves once no matter how many `share` has.

It seems that there should have similar optimization with an array-like index.
Anyway, consider this configuration:
```
{
     "chroot": "text$var",
     "share": [var1, var2, var3],
     "index": [var1, var2, var3, var4]
}
```

It's better not to resolve multiple times on the same variable.

Reply:
Do you think if we must support array-like index? Is a single index good enough for users?

Reply:
I don't know.  I use a single one myself.  We could maybe add support for string and variable, and hold array.  If someone ever requests it, we could merge the implementation we have, and if not, let's keep unit simpler.

I think it's rare to have a need for an array of indices.

Reply:
I'll remove the index-arr from this pull request, and keep it in a branch so that we have the code if we ever need it.

Reply:
Even a variable would be rarely needed for `index`, I would say.

Reply:
I'll post the optimization, so that you can check it.  Then I'll remove the array implementation from the PR.